README.md
## About
[Blurhash](https://github.com/woltapp/blurhash/) algorithm applied to sequence of images extracted from video file

List of hashes for 7 seconds video file with 16 frames per second has size of 6KB uncompressed!

## Preview
[Video preview](https://user-images.githubusercontent.com/21260939/209480508-4a372ae0-c4d5-4d92-82e8-305bea7838e4.mp4)

## Getting started

Include latest version from [pub.dev](https://pub.dev/packages/blurhash_video) to `pubspec.yaml`

## Usage

```dart
// generate sorted list of blurhashes from video 
final hashes = await BlurhashVideo.generateBlurHashes(
  path: path, // video file location 
  fps: 24, // video fps is used by default
  duration: 7, // in seconds
);
```


CHANGELOG.md
## 0.0.1

* Initial release

## 0.0.2

* Fix video preview

## 0.0.3

* Update minimal dependencies

## 0.0.4

* Implement clean up function
* Delete temp files on exception
* Format docstrings

## 0.0.5

* Disable scaling up for low-res videos
* Improve perfomance by disabling lossless compression and skipping multiple component calculations
* Update dependencies

## 0.0.6

* Fix multiple filters in ffmpeg command bug
* Fix input video fps upscaling
* Increase processing speed by setting compression_level to 50
* Option to set custom quality

## 0.0.7

* Update README.md

## 0.0.8

* Update package dependencies

lib/blurhash_video.dart
library blurhash_video;

import 'dart:io';
import 'dart:math';
import 'dart:collection';
import 'package:path/path.dart';
import 'package:path_provider/path_provider.dart';
import 'package:blurhash_dart/blurhash_dart.dart';
import 'package:image/image.dart';
import 'package:ffmpeg_kit_flutter_full_gpl/ffmpeg_kit.dart';
import 'package:ffmpeg_kit_flutter_full_gpl/return_code.dart';

/// Blurhash algorithm applied to video files using ffmpeg
class BlurhashVideo {
  /// Generate sequence of blurhashes from video file [path]
  /// * [workingDirectory] used for storing temporary image files, `getTemporaryDirectory()` used by default
  /// * [fps] used to specify how many hashes per second will be proceed, default to video fps
  /// * [duration] allows to cut video to specified lenght before processing, positive in seconds
  /// * [resolution] is widest side of thumbnail created from video, range from 32 to 64 pixels is enough because blurhash store just a bit of data after processing
  /// * [quality] is quality of PNG thumbnails, values in range `[0, 100]`
  static Future<List<String>> generateBlurHashes(
      {required String path,
      Directory? workingDirectory,
      int? fps,
      int? duration,
      int resolution = 64,
      int quality = 100}) async {
    SplayTreeMap<int, String> hashes = SplayTreeMap<int, String>();

    // temporary directory to save images
    Directory temp;
    if (workingDirectory != null) {
      temp = workingDirectory;
    } else {
      temp = await getTemporaryDirectory();
    }
    final timestamp = DateTime.now().millisecondsSinceEpoch;
    final destination = "${temp.path}/blurhash_video_temp_$timestamp";
    final directory = Directory(destination);
    await directory.create();

    // setup options
    final time = duration == null ? "" : "-t $duration";
    final frames = fps == null ? "" : "fps=fps='min(source_fps,$fps/1)',";
    final size =
        "scale=w='if(gte(iw,ih),min($resolution,iw),-2)':h='if(lt(iw,ih),min($resolution,ih),-2)'";
    // compression_level - size/speed tradeoff, 100 is smallest file and is default, 0 - bigger file while faster
    final q = "-quality $quality -compression_level 50 -pix_fmt rgb24";

    // run ffmpeg command
    final command =
        "-hide_banner -i \"$path\" $time -vf \"$frames$size\" $q \"$destination/%d.png\"";
    final session = await FFmpegKit.execute(command);

    final returnCode = await session.getReturnCode();
    if (!ReturnCode.isSuccess(returnCode)) {
      // remove temp directory and files
      await directory.delete(recursive: true);
      // produce ffmpeg output as exception
      throw "ffmpeg error: ${await session.getOutput()}";
    }

    // loop generated images
    int? numCompX, numCompY;
    await for (final file in directory.list()) {
      // skip system files & directories if any
      if (file is! File || !file.path.endsWith(".png")) continue;

      // extract id from 10.png format
      final id = basename(file.path).split(".").first;
      final index = int.parse(id);

      // read image
      final bytes = await file.readAsBytes();
      var image = decodePng(bytes);
      if (image == null) continue;

      // calculate components depending on resolution and rotation/orientation
      if (numCompX == null || numCompY == null) {
        final x = sqrt(16.0 * image.width / image.height);
        final y = x * image.height / image.width;
        numCompX = min(x.toInt() + 1, 9);
        numCompY = min(y.toInt() + 1, 9);
      }

      // generate hash
      final hash =
          BlurHash.encode(image, numCompX: numCompX, numCompY: numCompY);
      hashes[index] = hash.hash;
    }

    // clean up
    await directory.delete(recursive: true);

    return hashes.values.toList();
  }

  /// Delete all temporary directories and files created by this package
  /// May be used for clearing after application crashed
  /// * [workingDirectory] should be the same directory used for running [generateBlurHashes] function
  ///
  /// **Warning**: will delete all directories in [workingDirectory] starting with `blurhash_video_temp`
  static Future<void> cleanUp({
    Directory? workingDirectory,
  }) async {
    Directory temp;
    if (workingDirectory != null) {
      temp = workingDirectory;
    } else {
      temp = await getTemporaryDirectory();
    }

    // lookup directories in format "${temp.path}/blurhash_video_$timestamp"
    await for (final entry in temp.list()) {
      // filter files and wrong directories
      if (entry is! Directory ||
          !basename(entry.path).startsWith("blurhash_video_temp_")) continue;

      // delete the directory and all files in it
      await entry.delete(recursive: true);
    }
  }
}


example/lib/main.dart
import 'dart:io';
import 'package:ffmpeg_kit_flutter_full_gpl/ffmpeg_kit.dart';
import 'package:ffmpeg_kit_flutter_full_gpl/return_code.dart';
import 'package:flutter/material.dart';
import 'package:blurhash_video/blurhash_video.dart';
import 'package:file_picker/file_picker.dart';
import 'package:blurhash_dart/blurhash_dart.dart' as dart;
import 'package:image/image.dart';
import 'package:path_provider/path_provider.dart';
import 'package:image_sequence_animator/image_sequence_animator.dart';
import 'package:video_player/video_player.dart';

void main() {
  runApp(const App());
}

class App extends StatelessWidget {
  const App({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: "Blurhash Video Demo",
      theme: ThemeData(primarySwatch: Colors.deepOrange),
      home: const HomePage(),
      debugShowCheckedModeBanner: false,
    );
  }
}

class HomePage extends StatefulWidget {
  const HomePage({super.key});

  @override
  State<HomePage> createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  Directory? _directory;
  int _count = 0;

  VideoPlayerController? _gblurController;
  VideoPlayerController? _sequenceController;

  int _hashesSize = 0;
  int _gblurSize = 0;
  int _sequenceSize = 0;

  @override
  void dispose() async {
    super.dispose();
    _gblurController?.dispose();
    _sequenceController?.dispose();
    if (_directory?.existsSync() == true) {
      await _directory?.delete(recursive: true);
    }
  }

  void _process() async {
    // reset
    if (_directory?.existsSync() == true) {
      await _directory?.delete(recursive: true);
    }
    setState(() {
      _count = 0;
      _directory = null;
    });

    final temp = await getTemporaryDirectory();
    // print(temp.path);

    // clean all previously produced cache directories
    // await BlurhashVideo.cleanUp(workingDirectory: temp);

    final result = await FilePicker.platform.pickFiles(type: FileType.video);
    final path = result?.files.first.path;
    if (path == null) return;

    final hashes = await BlurhashVideo.generateBlurHashes(
      path: path,
      workingDirectory: temp,
      fps: 24, // video fps is used by default
      duration: 7, // in seconds
      resolution: 64, // pixels
      quality: 75, // from 0 to 100, bigger is better
    );

    // temporary directory for blurred images
    final dir = Directory(
        "${temp.path}/blurhash_video_demo_${DateTime.now().millisecondsSinceEpoch}");
    await dir.create();

    // make blurred image from each hash
    for (var i = 0; i < hashes.length; i++) {
      final blurHash = dart.BlurHash.decode(hashes[i]);
      final image = blurHash.toImage(64, 36);
      await File("${dir.path}/blur${i.toString().padLeft(5, "0")}.png")
          .writeAsBytes(encodePng(image));
    }

    setState(() {
      _count = hashes.length;
      _directory = dir;
    });

    // save hashes as txt file with new line for every next one
    final contents =
        hashes.fold<String>("", (content, item) => "$content$item\n");
    final hashesFile = File("${dir.path}/hashes.txt");
    await hashesFile.writeAsString(contents);
    // file can be compressed via bzip which will cut down to 1/3 of size
    final hashesSize = await hashesFile.length() ~/ 1024;
    setState(() => _hashesSize = hashesSize);

    // generate mp4 video with gaussian blur filter
    final gblurFile = File("${dir.path}/gblur.mp4");
    final gblurSession = await FFmpegKit.execute(
        "-hide_banner -i $path -s 64x36 -t 7 -vf gblur=sigma=128:steps=4 -lossless 1 -quality 100 -an ${gblurFile.path}");
    if (!ReturnCode.isSuccess(await gblurSession.getReturnCode())) {
      // print("ffmpeg error gblur: ${await gblurSession.getOutput()}");
      _gblurController?.dispose();
      _gblurController = null;
    } else {
      setState(() {
        _gblurController = VideoPlayerController.file(gblurFile)
          ..initialize().then((_) => _gblurController!.play());
      });
    }
    final gblurSize = await gblurFile.length() ~/ 1024;
    setState(() => _gblurSize = gblurSize);

    // generate mp4 video from blurred pngs
    final sequenceFile = File("${dir.path}/blurred.mp4");
    final sequenceSession = await FFmpegKit.execute(
        "-hide_banner -framerate 24 -i ${dir.path}/blur%05d.png -vf fps=24 -t 7 -pix_fmt yuv420p ${sequenceFile.path}");
    if (!ReturnCode.isSuccess(await sequenceSession.getReturnCode())) {
      // print("ffmpeg error sequence: ${await sequenceSession.getOutput()}");
      _sequenceController?.dispose();
      _sequenceController = null;
    } else {
      setState(() {
        _sequenceController = VideoPlayerController.file(sequenceFile)
          ..initialize().then((_) => _sequenceController!.play());
      });
    }
    final sequenceSize = await sequenceFile.length() ~/ 1024;
    setState(() => _sequenceSize = sequenceSize);
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text("Blurhash Video")),
      body: Column(
        mainAxisAlignment: MainAxisAlignment.spaceEvenly,
        children: [
          if (_directory != null) ...[
            // PNG Sequence
            ...[
              Text("PNG Sequence: $_hashesSize KB"),
              SizedBox(
                width: MediaQuery.of(context).size.width,
                height: MediaQuery.of(context).size.width * 0.5625,
                child: Transform.scale(
                  scale: MediaQuery.of(context).size.width / 64,
                  child: ImageSequenceAnimator(
                    _directory!.path,
                    "blur",
                    0,
                    5,
                    "png",
                    _count.toDouble(),
                    fps: 24,
                    isLooping: true,
                    isBoomerang: false,
                    isAutoPlay: true,
                  ),
                ),
              )
            ],

            // MP4 from image sequence
            if (_sequenceController?.value.isInitialized == true) ...[
              Text("MP4 Blurhash: $_sequenceSize KB"),
              AspectRatio(
                aspectRatio: _sequenceController!.value.aspectRatio,
                child: VideoPlayer(_sequenceController!),
              ),
            ],

            // MP4 ffmpeg blur effect
            if (_gblurController?.value.isInitialized == true) ...[
              Text("MP4 Gaussian ffmpeg: $_gblurSize KB"),
              AspectRatio(
                aspectRatio: _gblurController!.value.aspectRatio,
                child: VideoPlayer(_gblurController!),
              ),
            ]
          ] else
            const Center(
                child: Text("Select a video",
                    style:
                        TextStyle(fontWeight: FontWeight.w500, fontSize: 18))),
        ],
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _process,
        tooltip: "Select File",
        child: const Icon(Icons.file_present_rounded),
      ),
    );
  }
}
